File: golang/long-method/bad/user_manager.go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"fmt"
	"log"
	"regexp"
	"time"
)

type UserManager struct {
	db Database
}

func NewUserManager(db Database) *UserManager {
	return &UserManager{db: db}
}

func (um *UserManager) RegisterUser(userData map[string]string) (int64, error) {
	// Validate input data
	if userData["email"] == "" {
		return 0, fmt.Errorf("email is required")
	}
	if userData["password"] == "" {
		return 0, fmt.Errorf("password is required")
	}
	if len(userData["password"]) < 8 {
		return 0, fmt.Errorf("password must be at least 8 characters")
	}

	emailRegex := regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
	if !emailRegex.MatchString(userData["email"]) {
		return 0, fmt.Errorf("invalid email format")
	}

	// Check if user already exists
	rows, err := um.db.Query("SELECT id FROM users WHERE email = ?", userData["email"])
	if err != nil {
		return 0, err
	}
	defer rows.Close()

	if rows.Next() {
		return 0, fmt.Errorf("user already exists")
	}

	// Hash password
	hashedPassword := um.hashPassword(userData["password"])

	// Generate verification token
	verificationToken := um.generateToken()

	// Insert user into database
	result, err := um.db.Exec(`
		INSERT INTO users (email, password, first_name, last_name, verification_token, created_at)
		VALUES (?, ?, ?, ?, ?, ?)
	`, userData["email"], hashedPassword, userData["firstName"], userData["lastName"], verificationToken, time.Now())
	if err != nil {
		return 0, err
	}

	userId, err := result.LastInsertId()
	if err != nil {
		return 0, err
	}

	// Create user profile
	_, err = um.db.Exec(`
		INSERT INTO user_profiles (user_id, phone, address, city, state, zip_code)
		VALUES (?, ?, ?, ?, ?, ?)
	`, userId, userData["phone"], userData["address"], userData["city"], userData["state"], userData["zipCode"])
	if err != nil {
		return 0, err
	}

	// Send verification email
	subject := "Please verify your email address"
	message := fmt.Sprintf("Hello %s,\n\n", userData["firstName"])
	message += "Thank you for registering. Please click the link below to verify your email:\n\n"
	message += fmt.Sprintf("http://example.com/verify?token=%s\n\n", verificationToken)
	message += "Best regards,\nThe Team"

	fmt.Printf("Sending email to %s: %s\n", userData["email"], subject)
	// In real implementation, would send actual email

	// Log registration
	logMessage := fmt.Sprintf("User registered: %s at %s", userData["email"], time.Now().Format("2006-01-02 15:04:05"))
	log.Println(logMessage)

	// Create default settings
	_, err = um.db.Exec("INSERT INTO user_settings (user_id, theme, notifications_enabled) VALUES (?, 'light', true)", userId)
	if err != nil {
		return 0, err
	}

	// Send welcome notification
	_, err = um.db.Exec("INSERT INTO notifications (user_id, type, message, created_at) VALUES (?, 'welcome', 'Welcome to our platform!', ?)", userId, time.Now())
	if err != nil {
		return 0, err
	}

	return userId, nil
}

func (um *UserManager) hashPassword(password string) string {
	hash := sha256.Sum256([]byte(password))
	return hex.EncodeToString(hash[:])
}

func (um *UserManager) generateToken() string {
	bytes := make([]byte, 32)
	rand.Read(bytes)
	return hex.EncodeToString(bytes)
}

func main() {
	// This would normally use a real database connection
	var db Database

	um := NewUserManager(db)

	userData := map[string]string{
		"email":     "john@example.com",
		"password":  "password123",
		"firstName": "John",
		"lastName":  "Doe",
		"phone":     "555-1234",
		"address":   "123 Main St",
		"city":      "Anytown",
		"state":     "CA",
		"zipCode":   "12345",
	}

	userId, err := um.RegisterUser(userData)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}

	fmt.Printf("User registered with ID: %d\n", userId)
}

File: python/long-method/bad/user_manager.py
import hashlib
import secrets
import sqlite3
from datetime import datetime

class UserManager:
    def __init__(self, db):
        self.db = db

    def register_user(self, user_data):
        # Validate input data
        if not user_data.get('email'):
            raise Exception('Email is required')
        if not user_data.get('password'):
            raise Exception('Password is required')
        if len(user_data['password']) < 8:
            raise Exception('Password must be at least 8 characters')
        if not self._is_valid_email(user_data['email']):
            raise Exception('Invalid email format')

        # Check if user already exists
        cursor = self.db.cursor()
        cursor.execute("SELECT id FROM users WHERE email = ?", (user_data['email'],))
        if cursor.fetchone():
            raise Exception('User already exists')

        # Hash password
        hashed_password = self._hash_password(user_data['password'])

        # Generate verification token
        verification_token = secrets.token_hex(32)

        # Insert user into database
        cursor.execute(
            "INSERT INTO users (email, password, first_name, last_name, verification_token, created_at) VALUES (?, ?, ?, ?, ?, ?)",
            (
                user_data['email'],
                hashed_password,
                user_data['firstName'],
                user_data['lastName'],
                verification_token,
                datetime.now()
            )
        )

        user_id = cursor.lastrowid

        # Create user profile
        cursor.execute(
            "INSERT INTO user_profiles (user_id, phone, address, city, state, zip_code) VALUES (?, ?, ?, ?, ?, ?)",
            (
                user_id,
                user_data.get('phone'),
                user_data.get('address'),
                user_data.get('city'),
                user_data.get('state'),
                user_data.get('zipCode')
            )
        )

        # Send verification email
        subject = 'Please verify your email address'
        message = f"Hello {user_data['firstName']},\n\n"
        message += "Thank you for registering. Please click the link below to verify your email:\n\n"
        message += f"http://example.com/verify?token={verification_token}\n\n"
        message += "Best regards,\nThe Team"

        print(f"Sending email to {user_data['email']}")
        print(f"Subject: {subject}")
        print(f"Message: {message}")

        # Log registration
        log_message = f"User registered: {user_data['email']} at {datetime.now()}"
        print(f"LOG: {log_message}")

        # Create default settings
        cursor.execute(
            "INSERT INTO user_settings (user_id, theme, notifications_enabled) VALUES (?, 'light', 1)",
            (user_id,)
        )

        # Send welcome notification
        cursor.execute(
            "INSERT INTO notifications (user_id, type, message, created_at) VALUES (?, 'welcome', 'Welcome to our platform!', ?)",
            (user_id, datetime.now())
        )

        self.db.commit()
        return user_id

    def _is_valid_email(self, email):
        import re
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return re.match(pattern, email) is not None

    def _hash_password(self, password):
        return hashlib.sha256(password.encode()).hexdigest()
